Refactor Encoder class constructor to take needed options explicitly.

JianyuZhan opened the pull request:
#  **V1** 

This is the first patch series, addressing Issue https://github.com/OpenNMT/OpenNMT-py/issues/196.

This patch series refactors the `Encoder` class.

To do this, I chose a bottom-up way, that is refactoring the class used in `Encoder` class first, which are
`Embeddings` class and `TransformerEncoder` class, then refactor the `Encoder` class itself.

For the `Embeddings` class, since it is also used by `Decoder` class, which I will refactor in later series,
 I collect its needed options into a dictionary. So the `Encoder` class constructor would not be too cluttered.

To ease review, I split the patches into 4 parts incrementally:

patch 76fcdda2c5a2:  just a arg tidy-up.
patch 576d9cb560ac:  prep patch, refactored the `Embeddings` class
patch 49143d84ae20:  prep patch, refactored the `TransformerEncoder` class
patch dcf2a92882fa:  refactored the `Encoder` class

P.S. Every patch has been tested(flake8'ed, unittest'ed, run) successfully before applying the next one. 

P.P.S. These constructors might not be the final versions, but currently we just expose their needed options first, after these refactorization works are all finished, we could optimize them further if needed
(for example, the `cuda` option is used by many classes, so we can consider making it a gloal variable and not passing it).

JianyuZhan commented:
Travis CI tells some flake8 errors, and unittest errors.

I've fixed it up, and the patches' commit ID in the original pull request are also updated.

srush commented on onmt/Models.py:
```diff
@@ -548,17 +567,25 @@ def make_base_model(opt, model_opt, fields, cuda, checkpoint=None):
             break
         feature_vocabs.append(fields[key].vocab)
 
+    emb_opts = {'src_word_vec_size': model_opt.src_word_vec_size,
```
just have this make the embeddings.

srush commented on onmt/modules/Transformer.py:
```diff
@@ -46,15 +46,15 @@ def forward(self, x):
 
 class TransformerEncoder(nn.Module):
-    def __init__(self, hidden_size, opt, pad,
+    def __init__(self, hidden_size, dropout, pad,
```
This is the only place we need the padding index. See if you can get rid of it somehow.

JianyuZhan commented:
# Below is the **V2** message

This is the second iteration of this patch series. For the original message, see the section "Below is the V1 message" below.

## **Description**
This version has 6 patches, the **first  4 are the same with the V1**(with minor fix),
the **rest 2 are newly added**, mainly addressing the problems spotted in V1 review.

- Instead of passing `emb_opts`, `dicts` and `feature_dicts` into `Encoder`/`Decoder`
constructor, and making `Embedding` instance inside, now we just create it and directly
pass it to `Encoder` and `Decoder`. And I add a factory method for creating `Embeddings`.
Since sometimes just want to create an `Encoder` or `Decoder`(for example in test code), 
and don't bother to make `Embeddings` instance ourselves, then we can just use the `Embeddings` creation factory method.
- The `cuda` option could be obviated. Because `Embeddings` , `Encoder`
and `Decoder` are all subclasses of  `nn.Module`, so we can make `Embeddings` a sub-module
of `Encoder` and `Decoder`, thus it automatically inherits `Encoder`/`Decoder`'s `cuda` behavior.

## **Patches list**
--- below 4 are the same with V1, with minor errors fixed ---
patch 1f762c9c3b:  just a arg tidy-up.
patch 9c796efdca: prep patch, refactored the `Embeddings` class
patch 9047584fed:  prep patch, refactored the `TransformerEncoder` class
patch 485a39c80:  refactored the `Encoder` class

-- below 2 are new ---
patch 722f2a6c87: pass embeddings instance to Encoder/Decoder constructor.
patch d56cd70c33: get rid of cuda option

##  **Problems for dicussion**:

1. the pad_id: it is only used by `TransformerEncoder`, and we now still pass in into its constructor.
It is actually passed in to be used later by `get_attn_padding_mask`.  So I figure we can make `get_attn_padding_mask` to accepts a `dicts` argument. Thus `TransformerEncoder` would need
a `dicts` argument instead of a weird `pad_id`, which I think makes more sense. Since it is not related
what I am refactoring in this series, I just leave as is, and will address them later.
2. There is a really tricky problem when I use `add_module` to get rid of cuda option.  I quote my comment in the code: 

>   When we do  'python translate.py -model data/test_model.pt ...',
>   `model.load_state_dict(checkpoint['model'])` in `make_base_model`
>    will be called, which would compare the `state_dict.keys()` with
>    the loaded ones. However, `add_module()` would contribute one key 
>    to `state_dict.keys()`, thus making it have more keys than the 
>    loaded ones, and cause the load fail.
>
>    So we pretend we don't add one key by making it the same name
>    with the attribuite 'embeddings'.

srush commented on onmt/Models.py:
```diff
@@ -11,35 +11,33 @@
-    def __init__(self, vec_size, opt, dicts, feature_dicts=None):
+    def __init__(self, vec_size, position_encoding, feat_merge,
+                 feat_vec_exponent, feat_vec_size, dropout,
+                 dicts, feature_dicts=None):
```
Instead of passing the dictionaries, can you make this just pass in the sizes?

srush commented on onmt/Models.py:
```diff
@@ -120,48 +118,81 @@ def forward(self, src_input):
+            vec_size = opt.src_word_vec_size
+        else:
+            vec_size = opt.tgt_word_vec_size
+        return Embeddings(vec_size,
```
Can you pass these by keyword for readability.

srush commented on onmt/Models.py:
```diff
@@ -120,48 +118,81 @@ def forward(self, src_input):
     """
-    def __init__(self, opt, dicts, feature_dicts=None):
+    def __init__(self, encoder_type, brnn, rnn_type,
+                 num_layers, rnn_size, dropout, cuda,
```
I think you fixed the `cuda` issue here right?

srush commented:
Oops sorry, let me respond to you issues directly. Didn't see your comments, let me respond directly.

srush commented:
> 1) the pad_id: it is only used by TransformerEncoder, It is actually passed in to be used later by get_attn_padding_mask. So I figure we can make get_attn_padding_mask accept a dicts argument. Thus TransformerEncoder would need a dicts argument instead of a weird pad_id, which I think makes more sense. Since it is not related to what I am refactoring in this series, I just leave as is, and will address them later.

Yes, I think this can be done in a future PR. However, I would rather Transformer know *less* (just the pad_id) than more (the whole dict). Honestly, I don't think any of these classes should have access to the dictionary. The only two things they need are `len` for setting sizes and `pad_id` for masking. Whenever possible we should just give them those.

srush commented:
It looks like there is still a lot of CUDA being passed around. I don't really understand why that is necessary. Worse case we could override the `cuda()` function and do this, but I think the module should let us have constants.

srush commented:
> 2)  When we do test: 'python translate.py -model data/test_model.pt ...',
model.load_state_dict(checkpoint['model']) in make_base_model
will be called, which would compare the state_dict.keys() with
the loaded ones. However, add_module() would contribute one key
to state_dict.keys(), thus making it have one more key than the
loaded ones, and cause the load fail.

Thank you for the comment, but I don't exactly get the issue. Is the problem with loading old modules? Or is this a problem with ever passing in and attaching modules? Also not sure what this has to do with CUDA. 

Maybe you can look at what a class like RNN does when it is passed an nn module.

JianyuZhan commented:
>Thank you for the comment, but I don't exactly get the issue. Is the problem with loading old modules? >Or is this a problem with ever passing in and attaching modules? Also not sure what this has to do >with CUDA.
>
>Maybe you can look at what a class like RNN does when it is passed an nn module.

OK. Let me try to clarify this a little bit.

To get rid of cuda in `Embeddings`, I chose to call `Embed.add_module("embeddings", emb)`, so it get `cuda`  behavior as `Encoder` automatically. (The magic are all done in the `nn.Module` and `Embeddings` and `Encoder` are all subclass of it).

The side effort of `add_module()` is it contributes one key to `state_dict.keys()`, and which would be used to compare with the loaded ones when we load a checkpoint. And it will fail the load since the keys number now is unmatched.

srush commented:
Huh, do you need to call `add_module` though? I think that pytorch magically adds everything added after the constructor is called as a submodule. Can you check the semantics here?

bpopeters commented:
I'm not sure I understand the motivation for some of the changes to `Embeddings`. Why would one want to define it separate from the `Encoder` or `Decoder`? The provided reason is so that `Encoder` and `Decoder` can be tested without having to instantiate `Embeddings`, but I do not see that as a major advantage. An embedding matrix is an essential part of an `Encoder` or `Decoder` instance: neither of them works without it. Making the creation of the embedding matrix external makes the code less clear and more prone to bugs.

I am also a little skeptical about the `create_instance` static method. Creating an instance is the job of `__init__`: why should we additionally define a second function for it here? I can understand the desire to eliminate passing `opt` to lots of modules, but we haven't solved that problem if the method we actually use to instantiate that module takes `opt` anyway. Besides unpacking `opt`, the primary functionality implemented by `create_instance` is to pick between `opt.src_word_vec_size` and `opt.tgt_word_vec_size`. This is useful, but it will be unnecessary if `Embeddings` remains instantiated inside the `Encoder` and `Decoder` because whether to use the src or tgt size is clear from context.

srush commented:
I disagree with the first point. As long as there is a clear interface defined, I see no reason why decoder or encoder need to know anything about their embedding layers. By separating it out, we can test this interface to make sure each part is working, which I think makes it *less* prone to bugs. As people add different types of embeddings it is good that this is separate from encoder and decoder.

I partially agree with the second point. I think this function could be somewhere else (not on the object), and it would be fine, if you think that matters. My argument would be that before it was impossible to create an Embeddings object without knowing which opts it uses. Now you can, through __init__, which makes testing etc, possible. In that sense dependencies have been nicely separated.

bpopeters commented:
I can see where you're coming from about externally creating it. My view was that it's easy to enforce things like the input size of the RNN matching the embedding size if all of the submodules are defined inside `__init__`. However, I suppose it could be done elsewhere too. I'm just having memories of when I was implementing a model that required the dimension of `Decoder.forward()`'s `outputs` to be the hidden size multiplied by a constant factor, instead of just the hidden size. Hunting down the various places in the code where I subsequently had to alter the generator's dimensions was more difficult than necessary. I don't want make sure the Embeddings and RNN dimensions have matched dimensions to become like that.

I agree that it's nice for `Embeddings` to no longer take `opt` (and it would be nice for more things to follow). I made a similar change in #166, although instead of manipulating command line arguments in a static method inside an nn.Module subclass, I did it inside the Encoder/Decoder constructors. I don't think either approach is the right way to deal with command line arguments.

srush commented:
Okay, I think I understand the concern. Maybe let us keep on trying to separate. In the process we'll try to make it very clear the size of the input and output of each module, and even enforce by tests and assertions (so it makes clear right away if the sizes are wrong). And additionally we will explicitly have all the opt processing live outside the modules, as opposed to static in the class. 

This might take a couple iterations to get right, but I think we are mostly on the same page about getting rid of `opt`, not having the code become Java, with inheritance nonsense, and enforcing size matches.

JianyuZhan commented:
#  **V3** 

This is the 3rd iteration of this patch series. 

Compared to V2, major changes are:

1. Get rid of cuda from `Encoder`/`Decoder` class too. 
2. Make `Embeddings` constructor arguments resemble its superclass `nn.Module` as much as possible. This practice also is excerted on `Encoder` and later classes too. It is written as a guideline
into `Contributors.md`.  One small note:  some args thus would b not same name as the options listed
in `opts.py`, for example, `feat_vec_size` in `opts.py`, but it is called `feat_embedding_dim` in constructor to match the `embedding_dim` arg(previously called `vec_size`). But it is OK, we have a 
`Embeddings` factory method for creating instance, these details are hided.
3. Add a `tools/pull_request_chk.sh` script to do all sort of checks in one go(flake8, unittest, and tests listed in Travis.yml). `Contributors.md` is updated to guide future contributor to follow this before sending PR.
 
## **Patches list**

This has **7** patches, separated for easy review.

--- below **4** are **the same with V2**, with minor errors fixed ---
patch 1026520: refactored the `Embeddings` class.
patch e87c3cf:  refactored the `TransformerEncoder` class.
patch 3590f35:  refactored the `Encoder` class.

--- below **2** are the same with V2,  **with aforementioned major changes** ---
patch b4529c4: passed `Embeddings` instance to Encoder/Decoder constructor.
patch ba953d1: got rid of `cuda` from `Embeddings`.

--- below **2** are **new** in this iteration ---
patch 2d61ed3: got rid of `cuda` from `Encoder` and `Decoder`.
patch 0cfde37:  added `tools/pull_request_chk.sh` tool for easy check before sending PR.

srush commented on onmt/IO.py:
```diff
@@ -236,6 +236,18 @@ def collect_features(fields):
         return feats
 
     @staticmethod
+    def collect_feature_dicts(fields):
```
This is good. We'll fix this in a future PR (the torchtext people are working on it).

srush commented on onmt/Models.py:
```diff
@@ -4,59 +4,80 @@
+        feat_embedding_dim (int): embedding dimension for features when using
+                    '-feat_merge mlp'
+        dropout (float): dropout probablity.
+        dicts (Vocab): vocabulary dictionary.
```
Instead of passing a `dict` can you instead just pass its length and call it `num_word_embeddings` similary for the `features_dicts` . Whenever possible let's not pass `Vocab` objects around.

srush commented on onmt/Models.py:
```diff
@@ -113,55 +134,72 @@ def forward(self, src_input):
 
         return emb
 
+    @staticmethod
```
There was a request that this not live on the class object. I think that is fine, can you make it either of function or part of a different "controller" type class.

srush commented on onmt/Models.py:
```diff
@@ -113,55 +134,72 @@ def forward(self, src_input):
     """
-    def __init__(self, opt, dicts, feature_dicts=None):
+    def __init__(self, encoder_type, brnn, rnn_type,
+                 num_layers, rnn_size, dropout, enc_emb):
```
change `enc_emb` to `embeddings`

srush commented on onmt/Models.py:
```diff
@@ -113,55 +134,72 @@ def forward(self, src_input):
-            dicts (`Dict`): The src dictionary
-            features_dicts (`[Dict]`): List of src feature dictionaries.
+            encoder_type (string): rnn, brnn, mean, or transformer.
+            brnn (bool): brnn Encoder. Prefer `encoder_type=brnn` now.
```
change this to `bidirectional` to match pytorch.  drop the "prefer" part of the comment

srush commented on onmt/Models.py:
```diff
@@ -113,55 +134,72 @@ def forward(self, src_input):
-        self.hidden_size = opt.rnn_size // self.num_directions
-
+        # Call nn.Module.__init().
         super(Encoder, self).__init__()
```
Actually I think it is important that non-modules go before the constructor? Otherwise it saves them and breaks loading.

srush commented on onmt/Models.py:
```diff
@@ -215,7 +253,7 @@ class Decoder(nn.Module):
     """
 
-    def __init__(self, opt, dicts):
+    def __init__(self, opt, dec_emb):
```
`dec_emb` => `embeddings`

JianyuZhan commented:
#  **V4** 

This is the 4th iteration of this patch series. 

**V3 -> V4**, almost the same, with below changes added:

1. Now `Embeddings` accepts `dicts` and `feature_dicts`'s size(int) as argument, instead of `Vocab`. While at it, `padding_idx` is another arg passed to `Embeddings`,  because we now have no `Vocab` to retrieve it from. 
2. Make a standalone function `build_embeddings` for easy creation of `Embeddings`.
3. Args renaming:  `Encoder`: `brnn` -> `bidirectional`(following pytorch style),   `enc_emb` -> `embeddings`.  `Decoder`: `dec_emb` -> `embeddings`.
4. Rebased onto latest master branch.
 
## **Patches list**

This series has **7** patches, separated for easy review.
The aforementioned changes are mainly in patch 4.

1. patch 599dec6: refactored the `Embeddings` class.
2. patch 194cc04:  refactored the `TransformerEncoder` class.
3. patch 44bff54:  refactored the `Encoder` class.
4. patch 13bbee6: passed `Embeddings` instance to Encoder/Decoder constructor.
5. patch 8e7cdac: got rid of `cuda` from `Embeddings`.
6. patch 7ae5705: got rid of `cuda` from `Encoder` and `Decoder`.
7. patch 0bb72c1:  added `tools/pull_request_chk.sh` tool for easy check before sending PR.

bpopeters commented:
For what it's worth, I'm still doing the Embedding refactoring work I started in #166 . I've moved the class to its own space in onmt/modules and I'm doing a reasonable amount of refactoring in there to make the code clearer, as suggested by @srush . I think our work is mostly compatible and I'll try to incorporate your changes into mine when you're done.

JianyuZhan commented:
@bpopeters , Cool. Currently I am focusing on this `get rid of opt` re-factorization work and get familiar with the code at the same time. Glad it helps. Please go ahead.

srush commented:
Sounds great @bpopeters. I'll merge this and then you can compare to fixed code. Also maybe you can incorporate https://github.com/OpenNMT/OpenNMT-py/issues/210 into your PR. 

@JianyuZhan think this looks good, going to merge. 

Then onto `Decoder` next? Decoder might be a little more challenging. Think we may want to split  forward into 2 different functions (input feed and not), and split out DecoderTransformer into a different class.

JianyuZhan commented:
@srush , yep. I plan to refactor the `Decoder` next.  Okay, I will take all these into consideration.
